staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -352,6 +353,14 @@ func ValidateCustomResourceDefinitionOpenAPISchema(schema *apiextensions.JSONSch
 		allErrs = append(allErrs, ValidateCustomResourceDefinitionOpenAPISchema(schema.AdditionalProperties.Schema, fldPath.Child("additionalProperties"), ssv)...)
 	}
 
+	// CR validation will panic if pattern is not a valid regex.
+	// validate that the pattern is a valid regex before it is used for CR validation.
+	if len(schema.Pattern) != 0 {
```

Comment:
> What will happen if an invalid pattern is already saved in storage?

Good question. There are two cases possible here:

1. I have the CRD in storage and I have tried creating a CR before upgrading.

- In this case, the apiserver would have crashed when I created the CR, so having an invalid regex in storage then doesn't make much sense.

2. I have the CRD in storage but I haven't tried creating a CR before upgrading.

- In this case, I won't know that my regex is invalid - but when I would try creating a CR, it would crash my apiserver.
- If I upgrade -> create a CR -> apiserver still crashes: so no change in behaviour.
- If I upgrade -> go back and fix my CRD -> create a CR -> all ok.

@CaoShuFeng Wdyt?

Reply:
> If I upgrade -> go back and fix my CRD

tightening validation here can prevent an upgrade from succeeding. this seems like a use case for ratcheting validation... see https://github.com/kubernetes/kubernetes/pull/64907

I would also recommend translating the panic during custom resource validation into a validation failure instead of an API crash.

Reply:
> I would also recommend translating the panic during custom resource validation into a validation failure instead of an API crash.

Will create a PR upstream.

Reply:
This has been fixed upstream in https://github.com/go-openapi/validate/pull/42:
https://github.com/go-openapi/validate/blob/b0a3ed684d0fdd3e1eda00433382188ce8aa7169/values.go#L141-L150

```go
func Pattern(path, in, data, pattern string) *errors.Validation {
	re, err := compileRegexp(pattern)
	if err != nil {
		return errors.FailedPattern(path, in, fmt.Sprintf("%s, but pattern is invalid: %s", pattern, err.Error()))
	}
	if !re.MatchString(data) {
		return errors.FailedPattern(path, in, pattern)
	}
	return nil
}
```

I'll bump it in this PR instead then and wait for the ratcheting validation mechanism to land.

I'm not sure if we should cherry-pick the bump because afaik there have been other additions in go-openapi which we probably don't want in older versions (but would need to double check).

Reply:
Can'T we catch the panic for the older releases from our code?

Reply:
> Can'T we catch the panic for the older releases from our code?

Not sure I follow. Do you mean explicitly checking the version and then panicking according to that?

Reply:
```golang
defer func () {
  if recover() {
    return ...
  }
}
openapiValidate(...)
```

Reply:
Exactly. That's what I would want to do for older releases, transform that to an error, then ratchet the validation tighter in 1.12

The ratcheting PR is no longer WIP and is ready for review